jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +190,370 @@ default void clear()
+        }
+
+        public long capacityLong()
+        {
+            return _maxLength;
+        }
+
+        @Override
+        public boolean release()
+        {
+            clear();
+            return true;
+        }
+
+        @Override
+        public void clear()
+        {
+            for (RetainableByteBuffer buffer : _buffers)
+                buffer.release();
+            _canAggregate = false;
+            _buffers.clear();
+        }
+
+        @Override
+        public boolean append(ByteBuffer bytes)
+        {
+            int remaining = bytes.remaining();
+            int length = ensureMaxLength(remaining);
+            // if the length was restricted by maxLength, slice the bytes smaller
+            ByteBuffer out = bytes;
+            if (length < remaining)
+            {
+                ByteBuffer slice = bytes.slice();
+                slice.limit(slice.position() + length);
+                bytes.position(bytes.position() + length);
+                out = slice;
+            }
+            RetainableByteBuffer buffer = _pool.acquire(length, _direct);
+            buffer.append(out);
+            _buffers.add(buffer);
+            _canAggregate = true;
+            return !bytes.hasRemaining();
+        }
+
+        @Override
+        public boolean append(RetainableByteBuffer bytes)
+        {
+            int remaining = bytes.remaining();
+            int length = ensureMaxLength(remaining);
+
+            // If we cannot retain, then try aggregation into the last buffer
+            if (!bytes.canRetain() && _canAggregate)
+            {
+                if (_buffers.get(_buffers.size() - 1).append(bytes))
+                    return true;
+                length -= remaining - bytes.remaining();
+                remaining = bytes.remaining();
+            }
+
+            // if the length was restricted by maxLength, or we can't retain, then copy into new buffer
+            if (length < remaining || !bytes.canRetain())
+            {
+
+                RetainableByteBuffer buffer = _pool.acquire(length, _direct);
+                buffer.append(bytes);
+                bytes.clear();
+                _buffers.add(buffer);
+                _canAggregate = true;
+                return false;
+            }
+
+            bytes.retain();
+            _buffers.add(bytes);
+            _canAggregate = false;
+            return true;
+        }
+
+        @Override
+        public void putTo(ByteBuffer toInfillMode)
+        {
+            for (RetainableByteBuffer buffer : _buffers)
+                buffer.putTo(toInfillMode);
+        }
+
+        @Override
+        public void writeTo(Content.Sink sink, boolean last, Callback callback)
+        {
+            switch (_buffers.size())
+            {
+                case 0 -> callback.succeeded();
+                case 1 -> _buffers.get(0).writeTo(sink, last, callback);
+                default -> new IteratingCallback()
```

Comment:
Don't you need to succeed/fail the callback given as a parameter when this iterating callback finishes up?